Loading map menu

jmoerschbach opened the pull request:
Hi all,

this PR is not intended to be merged (yet), since it is still work in progress.
Some time soon, it will fix #3106 

However, I thought before adding new functionality, I'll do a cleanup/refactoring to improve code quality and readablity :)

I greatly appreciate feedback before proceeding (still, I'm a C++ novice, not really knowing what I'm doing... ;))

So, here is what I did:
* Separating 2 major responsibilities: reading/parsing files (LoadOrSaveGame) and displaying them (new class: SavegameTable)
* Introducing a small inheritance hierachy for the visualization to move further to the Single Responsibily Principle (no need to rely on boolean conditions all over the code)
* Extracting some functions in LoadOrSaveGame to make code more readable (there is still a lot of potential)

So, some questions from me:

* Is this more "object orientated" style welcome here? :)
* Is there anything terrible wrong (I'm thinking of memory leaks I am not aware of)?
* Why does LoadOrSaveGame not have a destructor in the first place?
 * Shouldn't at least UI::Box* table_box_ and UI::Button* delete_ be freed?

jmoerschbach committed:
extracted several methods to make big for loop readable

jmoerschbach committed:
introducing SavegameTable

jmoerschbach committed:
moved some more code, using SavegameTable now

jmoerschbach committed:
former functionality restored

jmoerschbach committed:
fix initial list

jmoerschbach committed:
stop refactoring, start implementing support for subdirectories now?

jmoerschbach committed:
introducing inheritance tree, now being able to finally work on #3106? :)

jmoerschbach committed:
set gametype before evaluating it, set sorting column, use suggested header guard

jmoerschbach committed:
Merge branch 'master' into loading_map_menu

jmoerschbach committed:
fix code style warnings

jmoerschbach committed:
fix includes & dependencies, still need to move map_filename (SavegameTable must not depend on filesystem)

jmoerschbach committed:
fix includes once again

jmoerschbach committed:
last time now?^^

stonerl commented:
You can run the code check locally. Have a look in the travis.sh file.

jmoerschbach committed:
moved includes, starting support for subdirectories now

jmoerschbach committed:
first draft of working subdirectories

jmoerschbach committed:
moved and cleaned up code

jmoerschbach commented:
Hi,

so I guess no feedback means nothing is terrible wrong, yes? :D

Sub-directories are now working, see screenshots:
![load menu](https://user-images.githubusercontent.com/10723431/69014186-f42c4b80-0987-11ea-9753-4b5f0a8b3cfb.png)
![load menu subdir](https://user-images.githubusercontent.com/10723431/69014196-0efec000-0988-11ea-848e-aa3e84149fc6.png)

What do you guys think about inserting directory/savegame icon (like in editor loading screen) to emphasize difference between file and directory?
![load map](https://user-images.githubusercontent.com/10723431/69014211-3786ba00-0988-11ea-9edc-4b2e0023cc32.png)

While there is still much refactoring potential, I think this might now be worth considering to merge...? :)

stonerl commented:
> What do you guys think about inserting directory/savegame icon (like in editor loading screen) to emphasize difference between file and directory?

+1

jmoerschbach committed:
remove unnecessary dependency

jmoerschbach committed:
extended table so clients can know number_of_columns, using this to remove redundant code, removed some unnecessary debug logs

jmoerschbach committed:
add icon support

jmoerschbach committed:
Merge branch 'master' into loading_map_menu

jmoerschbach committed:
Merge branch 'master' into loading_map_menu

jmoerschbach commented:
with icons:
![grafik](https://user-images.githubusercontent.com/10723431/69277351-7020d080-0be0-11ea-8867-529e6a205c9a.png)

I'll stop here now. There is still room for refactoring already existing code, but I think this should be merged first and then I may revisit this code again.

fixes #3106

@gunchleoc do you have time for review and feedback? :)

jmoerschbach commented:
any chance this will be merged sometime soon? :)

stonerl reviewed (dismissed):
LGTM from my side.

@tibor95 @Notabilis could one of you also have a brief look?

Notabilis reviewed (commented):
Thanks for the change, I like the feature! The code looks good overall and mostly it seems to work fine. I haven't really followed through which code has been refactored to where, but since it seems to work I guess you made no mistakes there. (Side note here: Doing the refactoring and the new feature in different branches makes it easier to review, but I end up with "mixed" branches often enough as well.)
I added a bunch of review comments to the diff of your changes. Most of them are only (not so important) code style changes, decide for yourself whether you want to address them.

Two things that should probably be changed:
- There seems to be a debug output with `log()` left, you probably intended to remove it?
- Two functions (`create_empty_dir()`, `create_directory()`) are declared but never implemented. I guess they can be removed?

One bug: The "Delete" button is broken for directories. Instead of printing the name of the directory to remove, it only displays an empty message. For savegames it displays the mapname, but of course that does not exist for directories.

I am unsure whether the sorting of the tables works as intended by you:
- Initial sorting after date works fine. If clicking the "Date" column again, the sorting is inverted so the parent directory ends up at the bottom. To fix this you probably have to modify the Table class, though. If you want to do so, maybe open up a new branch after this one has been merged.
- When sorting after the description/filename, the directories are no longer at the top but are sorted alphabetically somewhere in the middle. If you want to avoid this, I think you have to define a custom sort operator for this column as well.

And a feature request (for future branches): A "Create directory" option and maybe even a "Rename directory" option. ;-)

jmoerschbach commented:
Hi @Notabilis 
thank you very much for your profound review! I addressed the minor issues already.
Thanks for testing the delete button. I never tried it Oo
For the time being I think I will disable it when a directory is selected. Deleting a directory with savegames in it relates somehow to your requested features (creating/renaming directories), so I'll have another look to that on a new branch when this one is merged.

I'll fix the sorting of the description column though, since this looks really weird with the subdirectories now being visible

jmoerschbach committed:
addressed review issues:
made type_ private
made parameters and methods const where applicable
add curly brackets
removed unnecessary debug log output
removed method declaration which have no implementation

jmoerschbach committed:
moved compare methods
use dereference operator to get table entries
fix sorting of description column
made type_ private
<parent> instead of ..
disable delete button when directory is selected

jmoerschbach committed:
Merge branch 'master' of https://github.com/widelands/widelands into loading_map_menu

jmoerschbach committed:
Merge branch 'loading_map_menu' of https://github.com/jmoerschbach/widelands into loading_map_menu

jmoerschbach commented:
Hi,
I addressed all the issues from @Notabilis review. Ready for merging finally? :)

Notabilis commented:
Sorry, I found a new bug. :(
When selecting multiple files and directories, a list of them is shown on the right side of the screen. For directories, they are only listed as "," (since they don't have a mapname nor a save date). This is the same behavior as in the (now disabled) delete dialog. I think you have to modify `LoadOrSaveGame::filename_list_string` to fix this for the listings.

Some `table_->get(table_->get_record(` are still present, could you please replace them as well? Admittedly it is old code, but then everything will "look the same".

Apart from that, I am fine with the changes and the functionality, thanks!

jmoerschbach committed:
fixed display when selecting multiple files/directories

jmoerschbach commented:
Hi @Notabilis 
ok, next try ;)

thanks for testing! Hope everything is ok now.
Once this is merged I'll create another (refactoring only) branch since there are some issues I dislike, especially the misuse of SavegameData when multiple savegames are selected...

Notabilis commented:
Aaand back to you. Sorry! *duck and run*
It is looking good when selecting files and directories, but it crashes when only selecting multiple directories (but no files):
`#2  0x00007ffff640f40f in __assert_fail_base (fmt=0x7ffff6571ee0 "%s%s%s:%u: %s%sAssertion '%s' failed.\n%n", assertion=0x555557dad540 "filename.find('/') != std::string::npos",`
`    file=0x555557dad4e0 "src/wui/gamedetails.cc", line=214, function=<optimized out>) at assert.c:92`          
`#3  0x00007ffff641d102 in __GI___assert_fail (assertion=0x555557dad540 "filename.find('/') != std::string::npos", file=0x555557dad4e0 "src/wui/gamedetails.cc",`
`    line=214, function=0x555557daf5e0 <GameDetails::update(SavegameData const&)::__PRETTY_FUNCTION__> "void GameDetails::update(const SavegameData&)") at assert.c:101`
`#4  0x000055555760bf7a in GameDetails::update (this=0x7fffffff9480, gamedata=...) at src/wui/gamedetails.cc:214`
`#5  0x00005555576118c5 in LoadOrSaveGame::entry_selected (this=0x7fffffff9448) at src/wui/load_or_save_game.cc:162`
`#6  0x0000555556b4c98b in FullscreenMenuLoadGame::entry_selected (this=0x7fffffff8c60) at src/ui_fsmenu/loadgame.cc:165`

Suggestion: In wui/gamedetails.cc, change:
Line 186: `if (gamedata.filename_list.empty() && !gamedata.filename.empty()) {`
Line 236: `} else if (!gamedata.filename_list.empty()) {`
Not sure whether this is the best way to fix it, but at least it doesn't crash for me any longer.

Notabilis commented:
Just in case these questions are still open:

> Why does LoadOrSaveGame not have a destructor in the first place?
> Shouldn't at least UI::Box* table_box_ and UI::Button* delete_ be freed?

All UI elements are linked in a hierarchy of Panels (i.e., a rectangle that wants to be drawn). When creating a new Panel object (or one of its descendants, e.g., a button) a "higher" panel is given as parent in the constructor and the parent stores the newly created Panel as a child. When a Panel object is deleted, it also frees all it child Panels, so no explicit delete calls for the created buttons are required.
Consequently, we don't need a destructor in LoadOrSaveGame. The existing raw pointers are part of the panel hierarchy, so we don't have to care about deleting them. The other heap allocated objects are deleted by their unique_ptrs.

jmoerschbach committed:
fixed crash when only subdirectories are selected

jmoerschbach commented:
Applied your proposals for fixing the crash. This must be refactored, though to make it more clearer and readable in general... :)

Notabilis reviewed (dismissed):
Thanks. Not tested again, but looks like my own modification. As far as I am concerned, this can be merged now.
I think before merging you have to merge the current master into your branch (either by clicking the "Update branch" button or manually), clear up any potential conflicts, test again and of course the automatic checks have to succeed.

jmoerschbach committed:
Merge branch 'master' into loading_map_menu

gunchleoc reviewed (commented):
I apologize for not chiming in earlier, I had to take a break from the project because I became very busy with some other stuff.

Thanks for taking on this feature! I also like the refactoring.

I have added some nitpicky inline comments with code style suggestions.

Some additional observations: 

1. I noticed that the diff shows changes to debian/translations, which should not have any changes in it from this branch.
2. When taking on more refactoring later on, also have a look at mapselect.cc. There is some table sorting and display code for directories used in there somewhere. We might also be able to pull out some common code from SavegameData/MapData.

gunchleoc commented:
I did a very quick test and found a bug: Uncompressed savegames show up as directories and not as savegames.

jmoerschbach commented:
> 
> 
> I did a very quick test and found a bug: Uncompressed savegames show up as directories and not as savegames.

Oh, I wasn't aware of uncompressed savegames at all. I fixed that. Not sure if this is the most elegant way, though. What do you think. A possible alternative would be to try to actually load the savegame-directory and when that fails, treat it as a normal directory...

jmoerschbach committed:
adressed gunchleoc review issues
uncompressed savegames are now handled correctly

jmoerschbach committed:
Merge branch 'loading_map_menu' of https://github.com/jmoerschbach/widelands into loading_map_menu

gunchleoc commented:
Have a look at `FullscreenMenuMapSelect::fill_table()` on how uncompressed maps are detected. Maybe we can do something similar here?

jmoerschbach committed:
Merge branch 'master' into loading_map_menu

jmoerschbach committed:
*use raw pointer instead of unique_ptr
*use switch instead of if elseif
*try to load savegame, if it fails check if directory

jmoerschbach commented:
Hi @gunchleoc 
next try to address your review :)

FullscreenMenuMapSelect::fill_table() does something very similar to find out if it is a single file (by checking the filename's extension). If it is not considered as a regular file, it checks if it is a subdirectory and if so, adds a directory entry to the list.
The code here now tries to preload the savegame and if that fails, checks if it is a directory... Better?

gunchleoc commented:
Uncompressed save detection is working now on the load game screen :)

Found another bug though: When double-clicking on "example_folder" in the save game screen, Widelands saves the game to "example_folder.wfg" instead of entering the folder.

jmoerschbach committed:
saving a game in subdirectory works now, too

jmoerschbach commented:
> 
> 
> Uncompressed save detection is working now on the load game screen :)
> 
> Found another bug though: When double-clicking on "example_folder" in the save game screen, Widelands saves the game to "example_folder.wfg" instead of entering the folder.

fixed that

jmoerschbach committed:
Merge branch 'master' into loading_map_menu

gunchleoc commented:
All good now - merged :)
